sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -288,6 +288,9 @@ case class Join(
     condition: Option[Expression])
   extends BinaryNode with PredicateHelper {
 
+  /** Whether this join node is ordered before. Set true after it has been ordered. */
+  var ordered: Boolean = false
```

Comment:
Yeah your understanding is right, we could also memonize the joins we have already processed.